### PR TITLE
Transpose table iterator

### PR DESCRIPTION
**Pull request by drwells**
I tried to write some generic copying functions as part of #5956 but then realized that `TransposeTable` (the base class of `LAPACKFullMatrix`) does not implement an iterator interface. This PR adds a very nearly (I skipped `operator[]`) complete random access iterator interface for `TransposeTable`, based on (and largely compatible with) the one implemented for `SparseMatrix`: the accessors have `row()`, `column()`, and `value()` methods, work for const and non-const objects, are truly random access (no iterator increments involve loops).

I also simplified the `typedef`s for `TransposeTable` a bit. I am pretty sure that this is backwards compatible.

### CODE REVIEW

**Comment by drwells** [comment]
Update: I went back and fixed some typedefs, and made some other various improvements (I now store just 1 index instead of the row and column and fixed compatibility issues with what the standard says should be implemented).

**Comment by drwells** [comment]
I looked at this again, and I think with a bit more work we can use the same iterator template for all of the 2D array classes (which will get rid of a lot of duplicated code). I'll post again when this works.

**Comment by drwells** [comment]
This is now done.

It should now be possible to go back and implement complete iterators for `FullMatrix` and `SparseMatrix` with the new iterator template, but I will leave that for the future.

**Comment by drwells** [comment]
I know that this PR is not very exciting; I would appreciate if someone could look at it.

/run-tests

File: include/deal.II/base/linear_index_iterator.h [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+   */
+  bool operator == (const DerivedIterator &) const;
+
+  /**
+   * Inverse of <tt>==</tt>.
+   */
+  bool operator != (const DerivedIterator &) const;
+
+  /**
+   * Comparison operator: uses the same ordering as operator<, but also
```
Comment (masterleinad) [inline include/deal.II/base/linear_index_iterator.h]: operator<()

File: include/deal.II/base/linear_index_iterator.h [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+   * checks for equality.
+   *
+   * This function is only valid if both iterators point into the same
+   * container.
+   */
+  bool operator <= (const DerivedIterator &) const;
+
+  /**
+   * Comparison operator: uses the same ordering as operator>, but also
```
Comment (masterleinad) [inline include/deal.II/base/linear_index_iterator.h]: operator>()

File: include/deal.II/base/linear_index_iterator.h [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+ * Many classes in deal.II, such as FullMatrix, TransposeTable, and
+ * SparseMatrix, are fundamentally views into contiguous buffers: put another
+ * way, one can iterate across each of these classes with a random access
+ * iterator. This class is an iterator template, which can be used with any
+ * class that uses an array for its underlying storage. This class is
+ * essentially a simplified version of <code>boost::iterator_facade</code>
+ * that assumes <code>Accessor</code> provides certain features. The main
+ * intent of this class is that other classes can specify their own accessors
+ * and then use this class by inheriting with the curiously recurring template
+ * parameter (CRTP) technique, e.g.,
```
Comment (masterleinad) [inline include/deal.II/base/linear_index_iterator.h]: curiously recurring template pattern (CRTP).

The phrasing ("inhereting with the [...] technique" sounds a bit odd to me, but you are certainly more the expert than me.
I would have shortened this to "and then use the curiously recurring template pattern (CRTP) technique".

File: include/deal.II/base/linear_index_iterator.h [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+   */
+  typedef const value_type *pointer;
+
+  /**
+   * Size type used by the underlying container.
+   */
+  typedef typename value_type::size_type size_type;
+
+  /**
+   * copy operator.
```
Comment (masterleinad) [inline include/deal.II/base/linear_index_iterator.h]: Copy operator

File: include/deal.II/base/linear_index_iterator.h [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+  DerivedIterator &operator += (const difference_type n);
+
+  /**
+   * Fixed amount decrement.
+   */
+  DerivedIterator &operator -= (const difference_type n);
+
+  /**
+   * Return the distance between the current iterator and the argument. The
+   * distance is given by how many times one has to apply operator++ to the
```
Comment (masterleinad) [inline include/deal.II/base/linear_index_iterator.h]: operator++()

File: include/deal.II/base/linear_index_iterator.h [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+   * current iterator to get the argument (for a positive return value), or
+   * operator-- (for a negative return value).
```
Comment (masterleinad) [inline include/deal.II/base/linear_index_iterator.h]: operator--()

File: include/deal.II/base/linear_index_iterator.h [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+  pointer operator -> () const;
+
+  /**
+   * Comparison operator. Returns true if both iterators point to the same
+   * object position in the same object.
+   */
+  bool operator == (const DerivedIterator &) const;
+
+  /**
+   * Inverse of <tt>==</tt>.
```
Comment (masterleinad) [inline include/deal.II/base/linear_index_iterator.h]: or operator==()?

File: include/deal.II/base/table.h [hunk 8]
```diff
@@ -1499,6 +1501,204 @@ class Table<7,T> : public TableBase<7,T>
 
 
 
+/**
+ * A namespace for iterators for TransposeTable. TransposeTable is unique in
+ * that it stores entries in column-major (i.e., Fortran) order.
```
Comment (masterleinad) [inline include/deal.II/base/table.h]: I am not sure if this type of storage is a property unique to `TransposeTable`. Maybe rather "special"?

Do we really need to mention Fortran here?

File: tests/base/transpose_table_iterators_0.cc [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+template <typename Container>
+void test()
+{
+  Container table(3, 4);
+
+  auto it = table.begin();
+  auto t2 = it;
+  deallog << boost::core::demangle(typeid(Container).name()) << std::endl;
+  deallog << "++it == it:           " << (++table.begin() == table.begin()) << std::endl;
```
Comment (masterleinad) [inline tests/base/transpose_table_iterators_0.cc]: Would you mind using `std::boolalpha` here?

**Comment by drwells** [comment]
I implemented all suggestions.

File: include/deal.II/base/linear_index_iterator.h [hunk 10]
```diff
@@ -0,0 +1,449 @@
...
+ * @author David Wells, 2018
+ */
+template <class DerivedIterator, class AccessorType>
+class LinearIndexIterator
+{
+public:
+  /**
+   * Iterator category.
+   */
+  typedef std::random_access_iterator_tag iterator_category;
```
Comment (bangerth) [inline include/deal.II/base/linear_index_iterator.h]: Has C++11 deprecated deriving from `std::iterator` with this category? I found that clearer, given that inheritance denotes "is-a" relationships.

File: include/deal.II/base/linear_index_iterator.h [hunk 11]
```diff
@@ -0,0 +1,449 @@
...
+   * A typedef for the type you get when you dereference an iterator of the
+   * current kind.
+   */
+  typedef
+  AccessorType value_type;
+
+  /**
+   * Difference type.
+   */
+  typedef std::ptrdiff_t difference_type;
```
Comment (bangerth) [inline include/deal.II/base/linear_index_iterator.h]: You could probably generalize this if you wrote this as
```
  using difference_type = decltype(declval<Accessor>().linear_index - declval<Accessor>().linear_index);
```
i.e., the type of taking the difference between two `accessor.linear_index` numbers. This way, accessors don't need to use the large `ptrdiff_t` type if they don't want.

Not sure the obfuscation is worth the generalization, though.

File: include/deal.II/base/linear_index_iterator.h [hunk 12]
```diff
@@ -0,0 +1,449 @@
...
+}
+
+
+
+template <class DerivedIterator, class AccessorType>
+DerivedIterator
+LinearIndexIterator<DerivedIterator, AccessorType>::
+operator + (const difference_type n) const
+{
+  DerivedIterator copy(this->accessor);
```
Comment (bangerth) [inline include/deal.II/base/linear_index_iterator.h]: make it `const`

File: include/deal.II/base/linear_index_iterator.h [hunk 13]
```diff
@@ -0,0 +1,449 @@
...
+}
+
+
+
+template <class DerivedIterator, class AccessorType>
+DerivedIterator
+LinearIndexIterator<DerivedIterator, AccessorType>::
+operator - (const difference_type n) const
+{
+  DerivedIterator copy(this->accessor);
```
Comment (bangerth) [inline include/deal.II/base/linear_index_iterator.h]: same here

File: include/deal.II/base/linear_index_iterator.h [hunk 14]
```diff
@@ -0,0 +1,449 @@
...
+ * {
+ *   // implement additional constructors here, but all state should be contained
+ *   // in the Accessor.
+ * };
+ * @endcode
+ *
+ * The CRTP technique provides most of the boler-plate comparison operators
+ * for iterators. The implementation of this class assumes that
+ * <code>AccessorType</code> is a friend storing the offset
+ * <code>linear_index</code> and pointer <code>object</code>.
```
Comment (bangerth) [inline include/deal.II/base/linear_index_iterator.h]: Maybe link to some of the implementations of the uses of this class?

File: include/deal.II/base/table.h [hunk 15]
```diff
@@ -2367,6 +2624,170 @@ Table<2,T>::n_cols () const
...
+
+
+
+  template <typename T, bool Constness>
+  inline
+  AccessorBase<T, Constness>::AccessorBase(const object_pointer_type table)
+    :
+    object(table),
+    // TODO why can't this be object->values.size()?
+    linear_index(table->values.size())
```
Comment (bangerth) [inline include/deal.II/base/table.h]: That's because you should have gotten a warning that the compiler will reorder these two initializations: you declare the linear index *after* object in the class declaration, so even though you write it like this, the actual initialization happens the other way around.

In fact, I think that in all examples above (the sample doc example) and the actual `AccessorBase` class, you should declare `object` before `linear_index`: The latter is sub-ordinate to the former, so it makes sense to declare them in that order.

File: include/deal.II/base/table.h [hunk 16]
```diff
@@ -2367,6 +2624,170 @@ Table<2,T>::n_cols () const
...
+  {
+    Assert(!object->empty(), ExcMessage("An empty table has no rows or columns."));
+    return linear_index/object->n_rows();
+  }
+
+
+  template <typename T>
+  inline
+  T &
+  Accessor<T, false>::operator = (const T &t) const
```
Comment (bangerth) [inline include/deal.II/base/table.h]: I should have mentioned this at the class declaration: The common return type of `operator=` is the lhs operand, i.e., `*this`. This would also make the implementation of this function simpler.

Same for the `const`-version below.

File: include/deal.II/base/table.h [hunk 17]
```diff
@@ -2367,6 +2624,170 @@ Table<2,T>::n_cols () const
...
+    entry = t;
+    return entry;
+  }
+
+
+
+  template <typename T>
+  inline
+  T &
+  Accessor<T, false>::operator = (T &&t) const
```
Comment (bangerth) [inline include/deal.II/base/table.h]: same here

File: include/deal.II/base/linear_index_iterator.h (hunk 10 above)
Comment (drwells) [reply include/deal.II/base/linear_index_iterator.h]: If I recall correctly `std::iterator` was deprecated in C++17 (which is why I did not use it).

File: include/deal.II/base/table.h (hunk 15 above)
Comment (drwells) [reply include/deal.II/base/table.h]: That explains it. I am surprised I did not see compiler warnings.

**Comment by drwells** [comment]
@bangerth Thanks for the feedback. I believe we can use this to simplify some other code nicely (in particular, I find it frustrating that `FullMatrix` does not have a mutable iterator).

I originally implemented some of the index validity checking but I ultimately decided to remove it for two reasons:
1. The standard (or at least a copy of the STL specification that I have) says that `(i + n) - n == i` should work for any iterator `i` and integer `n`, so my understanding is that we should not do such range checks in iterators (though of course the accessor can).
2. It is convenient for count-down loops to permit an index of `-1` (which is why I used `std::ptrdiff_t`). If the iterator counts down to zero it ends with an (index) value of `-1`, which is convenient but would again not work with bounds checking.

File: include/deal.II/base/linear_index_iterator.h [hunk 18]
```diff
@@ -0,0 +1,449 @@
...
+#define dealii_linear_index_iterator_h
+
+#include <deal.II/base/config.h>
+#include <deal.II/base/exceptions.h>
+
+
+DEAL_II_NAMESPACE_OPEN
+/**
+ * Many classes in deal.II, such as FullMatrix, TransposeTable, and
+ * SparseMatrix, are fundamentally views into contiguous buffers: put another
```
[the remark this answers is not in the document]
Comment (drwells) [reply include/deal.II/base/linear_index_iterator.h]: This is much nicer than what I wrote. I used this text (with a little editing).

File: include/deal.II/base/linear_index_iterator.h [hunk 19]
```diff
@@ -0,0 +1,449 @@
...
+ *   // const iterators store a const pointer
+ *   typedef typename std::conditional<Constness, const FullMatrix<T>*, FullMatrix<T>*>::type
+ *   object_pointer_type;
+ *
+ * protected:
+ *   // The design of LinearIndexIterator assumes that these two members exist
+ *   // and completely describe the iterator's state. linear_index should have a
+ *   // signed type.
+ *   std::ptrdiff_t linear_index;
+ *   object_pointer_type object;
```
[the remark this answers is not in the document]
Comment (drwells) [reply include/deal.II/base/linear_index_iterator.h]: Hm. `parent_object` sounds like inheritance and `pointed_to_object` is a bit wordy, but `object` along is vague. How about `container`?

File: include/deal.II/base/linear_index_iterator.h [hunk 20]
```diff
@@ -0,0 +1,449 @@
...
+ *
+ *   // LinearIndexIterator needs access to linear_index and object.
+ *   friend class LinearIndexIterator<const_iterator<T>, Accessor<T, Constness>>;
+ *   friend class LinearIndexIterator<iterator<T>, Accessor<T, Constness>>;
```
[the remark this answers is not in the document]
Comment (drwells) [reply include/deal.II/base/linear_index_iterator.h]: They are a result of insufficient proof-reading. I redid this section without them.

File: include/deal.II/base/linear_index_iterator.h (hunk 12 above)
Comment (drwells) [reply include/deal.II/base/linear_index_iterator.h]: I am not sure that I understand this comment. I immediately increment `copy`; how could it be `const`?

File: include/deal.II/base/linear_index_iterator.h (hunk 13 above)
Comment (drwells) [reply include/deal.II/base/linear_index_iterator.h]: Same remark: I don't understand what you meant, since `copy` needs to be mutable.

File: include/deal.II/base/linear_index_iterator.h (hunk 14 above)
Comment (drwells) [reply include/deal.II/base/linear_index_iterator.h]: I added a note.

File: include/deal.II/base/table.h (hunk 16 above)
Comment (drwells) [reply include/deal.II/base/table.h]: Fixed.

File: include/deal.II/base/table.h (hunk 17 above)
Comment (drwells) [reply include/deal.II/base/table.h]: Fixed.

File: include/deal.II/base/linear_index_iterator.h (hunk 11 above)
Comment (drwells) [reply include/deal.II/base/linear_index_iterator.h]: I like the generality, but we do iterator differences so infrequently (at least for the matrix classes) that I don't think the optimization is worth making things more complicated.

**Comment by drwells** [comment]
I made most of the suggested changes: I think the technique for getting a difference type is too convoluted and I kept it as `ptrdiff_t`. I did essentially everthing else. Notes:

1. The hard part of this PR is the prose; I edited @bangerth's description of the class a bit but used most of it as-is.
2. I am still not sure what the suggested change is for `operator+` (and `operator-`).
3. I cleaned up the sample code.

File: include/deal.II/base/linear_index_iterator.h (hunk 12 above)
Comment (bangerth) [reply include/deal.II/base/linear_index_iterator.h]: My mistake. I was mentally still on `operator+=`.

File: include/deal.II/base/linear_index_iterator.h (hunk 10 above)
Comment (bangerth) [reply include/deal.II/base/linear_index_iterator.h]: You are correct of course: http://en.cppreference.com/w/cpp/iterator/iterator

File: include/deal.II/base/linear_index_iterator.h (hunk 19 above)
[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/base/linear_index_iterator.h]: Sure, `container` sounds reasonable. Also maybe `data_array` -- we know the memory is contiguous, so it's an array alright.

**Comment by bangerth** [comment]
On 03/16/2018 10:43 PM, David Wells wrote:
>  1. The standard (or at least a copy of the STL specification that I have)
>     says that |(i + n) - n == i| should work for any iterator |i| and integer
>     |n|, so my understanding is that we should not do such range checks in
>     iterators (though of course the accessor can).
>  2. It is convenient for count-down loops to permit an index of |-1| (which is
>     why I used |std::ptrdiff_t|). If the iterator counts down to zero it ends
>     with an (index) value of |-1|, which is convenient but would again not
>     work with bounds checking.

Fair enough. But can we at least add some kind of check when *dereferencing* 
such an iterator?

**Comment by drwells** [comment]
I would like to add this but I do not see how: we would need to write something like

```cpp
template <class DerivedIterator, class AccessorType>
inline
typename LinearIndexIterator<DerivedIterator, AccessorType>::reference
LinearIndexIterator<DerivedIterator, AccessorType>::operator * () const
{
  AssertIndexRange(accessor.linear_index, accessor.container->size());
  return accessor;
}
```

but not all relevant containers define a `size()` member function (e.g., `Table` and friends define `size(unsigned int)`. The unsatisfying answer is that the `Accessor` class already has some information about its container and it can figure out if it's index is valid or not; this functionality should belong there (or just rely on `AlignedVector::operator[]()`)

**Comment by bangerth** [comment]
What if instead of a pointer to an object you stored an `ArrayView`? It doesn't just have the pointer but also stores its length.

**Comment by drwells** [comment]
That would enable the bounds check, but we need a pointer to the object to get `row()` and `column()` in the accessor later on. Additionally, at some point I would like to write

```cpp
LAPACKBandedMatrix::begin(unsigned int band_n);
```

which would be a noncontiguous (but random access) iterator over a certain band: `ArrayView` won't work here.

**Comment by bangerth** [comment]
Ah, I see how you're running into the problem now that the iterator class needs to know something about the accessor class. You can't just implement the accessor however you want :-( Any other idea?

**Comment by drwells** [comment]
One issue with this template is that it needs to be general enough to work with all our matrix-like containers. Part of that means surrendering some logic to the `Accessor` class. I don't see any other way (besides what I posted above, i.e., make the `Accessor` check everything) to do range checking for this reason.

I updated the documentation some more; these are useful questions and hopefully things are now clearer.

File: include/deal.II/base/linear_index_iterator.h [hunk 21]
```diff
@@ -0,0 +1,499 @@
...
+ * <code>AccessorType</code> provides certain members (documented below) that
+ * completely describe the state of the iterator. The intended use of this
+ * class is for containers to define their own accessor classes and then use
+ * the curiously recurring template pattern (CRTP) technique to define their
+ * iterators, e.g.,
+ *
+ * @code
+ * // forward declaration for pointer definition
+ * template <typename T>
+ * class Container;
```
Comment (bangerth) [inline include/deal.II/base/linear_index_iterator.h]: can you say somewhere that `Container<T>` is in fact the container into which we want to have iterators?

Is it important that it actually has a template argument?

File: include/deal.II/base/linear_index_iterator.h [hunk 22]
```diff
@@ -0,0 +1,499 @@
...
+ *
+ * // forward declaration for friendship
+ * template <typename T, bool Constness>
+ * class Iterator;
```
Comment (bangerth) [inline include/deal.II/base/linear_index_iterator.h]: What is `T` here? If we could get rid of the template argument of the container, then maybe we can drop it here as well?

File: include/deal.II/base/linear_index_iterator.h [hunk 23]
```diff
@@ -0,0 +1,499 @@
...
+ *
+ * template <typename T, bool Constness>
+ * class Accessor
```
Comment (bangerth) [inline include/deal.II/base/linear_index_iterator.h]: Maybe call these classes `ContainerAccessor` and `ContainerIterator` (below) to make clear that these are not just any generic classes, but are specifically for the `Container` class.

Or maybe even call the class `MyContainer` and `MyContainerAccessor/Iterator` to make it even more clear that these are specific classes?

File: include/deal.II/base/linear_index_iterator.h [hunk 24]
```diff
@@ -0,0 +1,499 @@
...
+ * boiler-plate comparison and arithmetic operators for iterators. This is
+ * necessary for, e.g., LinearIndexIterator::operator++() to return the
+ * correct type.
+ *
+ * @tparam AccessorType LinearIndexIterator assumes that the
+ * <code>AccessorType</code> template parameter has the following members
+ * which completely describe the current state of the iterator:
+ * <ol>
+ *   <li>A pointer named <code>container</code> to the original container (e.g.,
+ *   the *   relevant SparseMatrix). This should be a <code>const</code> pointer
```
Comment (bangerth) [inline include/deal.II/base/linear_index_iterator.h]: messed up formatting here

File: include/deal.II/base/table.h [hunk 25]
```diff
@@ -2367,6 +2624,168 @@ Table<2,T>::n_cols () const
...
+
+
+  template <typename T, bool Constness>
+  inline
+  const T &
+  AccessorBase<T, Constness>::value() const
+  {
+    Assert(0 <= this->linear_index,
+           ExcMessage("The current iterator points outside of the table."));
+    return this->container->values[this->linear_index];
```
Comment (bangerth) [inline include/deal.II/base/table.h]: Don't you here know the type of the container and could ensure that the linear index is valid?

File: include/deal.II/base/table.h [hunk 26]
```diff
@@ -2367,6 +2624,168 @@ Table<2,T>::n_cols () const
...
+
+
+
+  template <typename T, bool Constness>
+  inline
+  typename AccessorBase<T, Constness>::size_type
+  AccessorBase<T, Constness>::AccessorBase::row() const
+  {
+    Assert(!container->empty(), ExcMessage("An empty table has no rows or columns."));
+    return linear_index % container->n_rows();
```
Comment (bangerth) [inline include/deal.II/base/table.h]: same here and in all other places, obviously -- I mean, you do access `n_rows()`, so couldn't you also make sure that `linear_index` is within range?

File: include/deal.II/base/table.h [hunk 27]
```diff
@@ -2367,6 +2624,168 @@ Table<2,T>::n_cols () const
...
+  {}
+
+
+
+  template <typename T, bool Constness>
+  inline
+  AccessorBase<T, Constness>::AccessorBase(const container_pointer_type table)
+    :
+    container(table),
+    linear_index(container->values.size())
```
Comment (bangerth) [inline include/deal.II/base/table.h]: Do you think of this as the end iterator then?

File: include/deal.II/base/table.h [hunk 28]
```diff
@@ -2367,6 +2624,168 @@ Table<2,T>::n_cols () const
...
+
+
+  template <typename T, bool Constness>
+  inline
+  AccessorBase<T, Constness>::AccessorBase(const container_pointer_type table,
+                                           const std::ptrdiff_t         index)
+    :
+    container(table),
+    linear_index(index)
+  {}
```
Comment (bangerth) [inline include/deal.II/base/table.h]: I think you could check for validity of the index here, i.e., whether it is in range for the container.

File: include/deal.II/base/table.h (hunk 25 above)
Comment (drwells) [reply include/deal.II/base/table.h]: We can add the check. I skipped it since it is technically redundant (`AlignedVector` will check) but I will add it back in.

File: include/deal.II/base/table.h (hunk 27 above)
Comment (drwells) [reply include/deal.II/base/table.h]: Yes (see the constructor documentation :))

File: include/deal.II/base/table.h (hunk 26 above)
Comment (drwells) [reply include/deal.II/base/table.h]: Good point. I'll add a check.

File: include/deal.II/base/table.h (hunk 28 above)
Comment (drwells) [reply include/deal.II/base/table.h]: Yes; I will add this.

File: include/deal.II/base/linear_index_iterator.h (hunk 24 above)
Comment (drwells) [reply include/deal.II/base/linear_index_iterator.h]: Fixed.

File: include/deal.II/base/linear_index_iterator.h (hunk 23 above)
Comment (drwells) [reply include/deal.II/base/linear_index_iterator.h]: I interpreted your second suggestion as 'implement the iterator and accessor as classes inside `Container`' and did it.

File: include/deal.II/base/linear_index_iterator.h (hunk 22 above)
Comment (drwells) [reply include/deal.II/base/linear_index_iterator.h]: The template argument is not significant; I added `const T& value() const` to the Accessor to make it clearer where this is used. I think the arbitrary-ness of `T` is helpful when describing a container (especially if we actually use it).

File: include/deal.II/base/table.h (hunk 25 above)
Comment (bangerth) [reply include/deal.II/base/table.h]: Oh, I had not realized that. I thought that `values` is just some `std::vector` or `double*` or something.

But it probably doesn't hurt either to check here already, rather than further down.

File: include/deal.II/base/linear_index_iterator.h (hunk 23 above)
Comment (bangerth) [reply include/deal.II/base/linear_index_iterator.h]: That works too!

**Comment by bangerth** [comment]
Let me know when you want me to take another look!

**Comment by drwells** [comment]
@bangerth thanks :) I added the range checks and updated the example.

**Comment by bangerth** [comment]
Great, thanks for going through the iterations with me! (Pun intended.)

OK to merge once the tester is happy.

File: include/deal.II/base/table.h [hunk 29]
```diff
@@ -2367,6 +2624,177 @@ Table<2,T>::n_cols () const
...
+
+
+  template <typename T, bool Constness>
+  inline
+  typename AccessorBase<T, Constness>::size_type
+  AccessorBase<T, Constness>::AccessorBase::row() const
+  {
+    Assert(!container->empty(), ExcMessage("An empty table has no rows or columns."));
+    const auto row_n = linear_index % container->n_rows();
+    Assert(0 <= row_n && row_n < container->m(),
```
Comment (masterleinad) [inline include/deal.II/base/table.h]: The failing tests are suggesting that this should be `container->n_rows()` (or `TransposeTable` should implement `m()`).

File: include/deal.II/base/table.h [hunk 30]
```diff
@@ -2367,6 +2624,177 @@ Table<2,T>::n_cols () const
...
+
+
+  template <typename T, bool Constness>
+  inline
+  typename AccessorBase<T, Constness>::size_type
+  AccessorBase<T, Constness>::AccessorBase::column() const
+  {
+    Assert(!container->empty(), ExcMessage("An empty table has no rows or columns."));
+    const auto column_n = linear_index/container->n_rows();
+    Assert(0 <= column_n && column_n < container->n(),
```
Comment (masterleinad) [inline include/deal.II/base/table.h]: The failing tests are suggesting that this should be `container->n_columns()` (or `TransposeTable` should implement `n()`).

**Comment by drwells** [comment]
The failing test is due to a wrong bounds check: we cannot write
```cpp
Assert(0 <= linear_index && linear_index < container->values.size(), ExcMessage("..."));
```

since this prevents us from creating the end iterator (or the begin iterator of an empty table). I fixed it by writing
```cpp
Assert(0 <= linear_index && linear_index < container->values.size() + 1, ExcMessage("..."));
```

that should fix the failing test.
